### PR TITLE
ISPN-13497 IRAC Batch keys in update request

### PR DESCRIPTION
**Pull request by pruivo**
Instead of sending each key individually, batch them in a single
request.

https://issues.redhat.com/browse/ISPN-13497

### CODE REVIEW

File: core/src/main/java/org/infinispan/xsite/irac/IracResponseCollector.java [hunk 1]
```diff
@@ -12,83 +13,117 @@
 import org.infinispan.util.logging.LogFactory;
 import org.infinispan.xsite.status.DefaultTakeOfflineManager;
 
+import net.jcip.annotations.GuardedBy;
+
 /**
  * A response collector for a single key update.
  * <p>
  * This class extends {@link CompletableFuture}. It is completed when all sites involved in the updated have replied (or
```
Comment (tristantarrant) [inline core/src/main/java/org/infinispan/xsite/irac/IracResponseCollector.java:21]: The javadoc is lying: it extends `Runnable`

Comment (tristantarrant) [reply core/src/main/java/org/infinispan/xsite/irac/IracResponseCollector.java:21]: I think it is also not about single keys anymore

**Review by tristantarrant — APPROVED** [review]
LGTM

**Comment by tristantarrant** [comment]
Do we have perf tests ?

**Comment by pruivo** [comment]
> Do we have perf tests ?

nothing recently; It was slow (<5% if my memory isn't failing) because `Flowable` waits for the previous batch to be acked before sending the next one.

File: core/src/main/java/org/infinispan/xsite/irac/IracResponseCollector.java (hunk 1 above)
Comment (pruivo) [reply core/src/main/java/org/infinispan/xsite/irac/IracResponseCollector.java]: I've updated the javadoc.

**Comment by pruivo** [comment]
latest results:
Baseline (cc12b18106ce58706f2cc7f3bcb83a8b75483030): 38789 reqs/sec
PR: 50502 reqs/sec

**Comment by tristantarrant** [comment]
Ok, that's a big improvement

**Comment by wburns** [comment]
> latest results: Baseline ([cc12b18](https://github.com/infinispan/infinispan/commit/cc12b18106ce58706f2cc7f3bcb83a8b75483030)): 38789 reqs/sec PR: 50502 reqs/sec

Very awesome! I have been waiting for this PR for a while when I saw some of the other xsite changes 💯

File: core/src/main/java/org/infinispan/xsite/irac/DefaultIracManager.java [hunk 2]
```diff
@@ -342,54 +355,23 @@ private void trackState(IracManagerKeyState state) {
...
-      return CompletableFutures.completedNull();
+      CompletableFutureObservable cfo = new CompletableFutureObservable();
+      Flowable.fromIterable(updatedKeys.values())
+            .filter(this::canStateBeSent)
+            .concatMapSingle(this::fetchEntry)
+            .filter(Objects::nonNull)
+            .buffer(batchSize)
+            .concatMapCompletable(this::sendUpdateBatch)
+            .subscribe(cfo);
+      return cfo;
```
Comment (wburns) [inline core/src/main/java/org/infinispan/xsite/irac/DefaultIracManager.java:374]:
```suggestion
      return Flowable.fromIterable(updatedKeys.values())
            .filter(this::canStateBeSent)
            .concatMapMaybe(this::fetchEntry)
            .buffer(batchSize)
            .concatMapCompletable(this::sendUpdateBatch)
            .onErrorComplete(t -> {
               onUnexpectedThrowable(t);
               return true;
            })
            .toCompletionStage(null);
```
And remove the observable class at the end of the file.

Note this also requires the changes to change from Single to Maybe for fetchEntry.

File: core/src/main/java/org/infinispan/xsite/irac/DefaultIracManager.java [hunk 3]
```diff
@@ -400,41 +382,127 @@ public boolean isEmpty() {
...
+         // topology changed! we no longer have the key; just drop it
+         state.discard();
+         removeStateFromLocal(state);
+         return false;
+      }
+      return dInfo.isPrimary() && state.canSend();
+   }
+
+   private Single<IracStateData> fetchEntry(IracManagerKeyState state) {
+      return Single.fromCompletionStage(clusteringDependentLogic.getEntryLoader()
```
Comment (wburns) [inline core/src/main/java/org/infinispan/xsite/irac/DefaultIracManager.java:399]:
```suggestion
   private Maybe<IracStateData> fetchEntry(IracManagerKeyState state) {
      return Maybe.fromCompletionStage(clusteringDependentLogic.getEntryLoader()
```

Comment (wburns) [reply core/src/main/java/org/infinispan/xsite/irac/DefaultIracManager.java:399]: Requires the change above to use `concatMapMaybe` as well

File: core/src/main/java/org/infinispan/xsite/irac/DefaultIracManager.java [hunk 4]
```diff
@@ -400,41 +382,127 @@ public boolean isEmpty() {
...
+      if (invalidState.isEmpty()) {
+         return Completable.fromCompletionStage(rspCollector.freeze());
+      }
+
+      if (trace) {
+         log.tracef("[IRAC] Removing %d invalid state(s)", invalidState.size());
+      }
+      invalidState.forEach(IracManagerKeyState::discard);
+      removeStateFromCluster(invalidState);
+      return Completable.fromCompletionStage(rspCollector.freeze());
```
Comment (wburns) [inline core/src/main/java/org/infinispan/xsite/irac/DefaultIracManager.java:468]:
```suggestion
      if (!invalidState.isEmpty()) {
         if (trace) {
            log.tracef("[IRAC] Removing %d invalid state(s)", invalidState.size());
         }
         invalidState.forEach(IracManagerKeyState::discard);
         removeStateFromCluster(invalidState);
      }

      return Completable.fromCompletionStage(rspCollector.freeze());
```
Just a bit more compact.

File: core/src/main/java/org/infinispan/xsite/irac/IracClearResponseCollector.java [hunk 5]
```diff
@@ -0,0 +1,78 @@
...
+               //don't overwrite communication errors
+               RESULT_UPDATED.compareAndSet(this, IracBatchSendResult.OK, IracBatchSendResult.RETRY);
+            }
+            if (backup.logExceptions()) {
+               log.warnXsiteBackupFailed(cacheName, backup.getSiteName(), throwable);
+            } else if (trace) {
+               log.tracef(throwable, "Encountered issues while backing clear command for cache %s to site %s", cacheName, backup.getSiteName());
+            }
+         } else if (trace) {
+            log.tracef("Received clear response from %s (%d missing)", backup.getSiteName(), countDownRunnable.missing());
```
Comment (wburns) [inline core/src/main/java/org/infinispan/xsite/irac/IracClearResponseCollector.java:65]:
```suggestion
            log.tracef("Received clear response from %s (%d remaining)", backup.getSiteName(), countDownRunnable.missing());
```
Missing makes it sound like something is wrong.

**Review by wburns — COMMENTED** [review]
Looks good to me, just some minor things. However, I would recommend using IntSet instead of BitSet in places.

File: commons/all/src/main/java/org/infinispan/commons/marshall/Ids.java [hunk 6]
```diff
@@ -225,6 +225,8 @@
    Integer XSITE_AUTO_TRANSFER_RESPONSE = 153;
    Integer COMMAND_INVOCATION_ID = 154;
 
+   Integer BIT_SET = 155;
```
[the remark this answers is not in the document]
Comment (pruivo) [reply commons/all/src/main/java/org/infinispan/commons/marshall/Ids.java:228]: sure

**Comment by pruivo** [comment]
updated. let me know when I can squash the commits :+1:

File: core/src/main/java/org/infinispan/commands/irac/IracPutManyCommand.java [hunk 7]
```diff
@@ -0,0 +1,241 @@
...
+      super(COMMAND_ID, cacheName);
+   }
+
+   public IracPutManyCommand(ByteString cacheName, int maxCapacity) {
+      super(COMMAND_ID, cacheName);
+      updateList = new ArrayList<>(maxCapacity);
+   }
+
+   public CompletionStage<IntSet> executeOperation(BackupReceiver receiver) {
+      IntSet rsp = IntSets.mutableEmptySet(updateList.size());
```
Comment (wburns) [inline core/src/main/java/org/infinispan/commands/irac/IracPutManyCommand.java:63]: I would just use a concurrent int set. Then you don't need to synchronize below as well.

**Review by wburns — COMMENTED** [review]
Looks good, you can squash. Last comment is really minor.

**Review by wburns — APPROVED** [review]
LGTM, if you squash I can merge.

**Comment by pruivo** [comment]
Failures are not related. It is a problem with the test itself; details in https://issues.redhat.com/browse/ISPN-13768

**Comment by tristantarrant** [comment]
Merged, thanks

**Pull request merged**